pkg/plugin/processor/builtin/impl/ai/openai/embedding.go

```diff
@@ -0,0 +1,162 @@
+
+	inputFieldRefResolver sdk.ReferenceResolver
+
+	client *azopenai.Client
+	logger log.CtxLogger
+}
+
+func NewEmbeddingProcessor(log log.CtxLogger) sdk.Processor {
+	return &embeddingProcessor{
+		logger: log.WithComponent("openai_embedding"),
+	}
+}
+
+func (p *embeddingProcessor) Specification() (sdk.Specification, error) {
+	return sdk.Specification{
+		Name:        "openai.embedding",
+		Summary:     "Generate OpenAI embeddings.",
+		Description: "",
+		Version:     "v0.1.0",
+		Author:      "Meroxa, Inc.",
+		Parameters:  embeddingProcConfig{}.Parameters(),
+	}, nil
+}
+
+func (p *embeddingProcessor) Configure(ctx context.Context, c config.Config) error {
+	cfg := embeddingProcConfig{}
+	err := sdk.ParseConfig(ctx, c, &cfg, embeddingProcConfig{}.Parameters())
+	if err != nil {
+		return cerrors.Errorf("failed to parse configuration: %w", err)
+	}
+
+	inputResolver, err := sdk.NewReferenceResolver(cfg.InputField)
+	if err != nil {
+		return cerrors.Errorf(`failed to create a field resolver for %v parameter: %w`, cfg.InputField, err)
+	}
+	p.inputFieldRefResolver = inputResolver
+
+	p.cfg = cfg
+	return nil
+}
+
+func (p *embeddingProcessor) Open(ctx context.Context) error {
+	keyCredential := azcore.NewKeyCredential(p.cfg.APIKey)
+
+	// NOTE: this constructor creates a client that connects to the public OpenAI endpoint.
+	// To connect to an Azure OpenAI endpoint, use azopenai.NewClient() or azopenai.NewClientWithyKeyCredential.
+	client, err := azopenai.NewClientForOpenAI("https://api.openai.com/v1", keyCredential, nil)
+	if err != nil {
+		return cerrors.Errorf("failed to create OpenAI client: %w", err)
+	}
+
+	p.client = client
+	return nil
+}
+
+func (p *embeddingProcessor) Process(ctx context.Context, records []opencdc.Record) []sdk.ProcessedRecord {
+	out := make([]sdk.ProcessedRecord, 0, len(records))
+
+	// Prepare request (batch all embedding inputs)
+	var embeddingInputs []string
+	for _, record := range records {
+		inRef, err := p.inputFieldRefResolver.Resolve(&record)
+		if err != nil {
+			out = append(out, sdk.ErrorRecord{Error: fmt.Errorf("failed to resolve reference %v: %w", p.cfg.InputField, err)})
+			continue
+		}
+
+		embeddingInputs = append(embeddingInputs, p.getEmbeddingInput(inRef.Get()))
+	}
+
+	// Execute request (get embeddings)
+	embeddings, err := p.client.GetEmbeddings(
+		ctx,
+		azopenai.EmbeddingsOptions{
+			Input:          embeddingInputs,
+			EncodingFormat: lang.Ptr(azopenai.EmbeddingEncodingFormatBase64),
+			DeploymentName: &p.cfg.Model,
+		},
+		nil,
+	)
+	// If the request failed, declare processing for all records as failed
+	if err != nil {
+		for range len(records) {
+			out = append(out, sdk.ErrorRecord{Error: err})
+		}
+
+		return out
+	}
+
+	p.logger.Trace(ctx).
+		Any("embedding_input", embeddingInputs).
+		Any("embedding_output", embeddings).
+		Msg("got embeddings")
+
+	for i, record := range records {
+		record.Metadata[EmbeddingMetadataBase64] = embeddings.Data[i].EmbeddingBase64
```

Comment:
These can be horribly large, my suggestion here is to:

1. add the embedding to `.Payload.After`
2. include the model name used for the embedding.

Particularly 2, since there are others openai compat systems which may allow for different models to be used.

Reply:
You're right, especially about no. 2. As for 1, that would also mean that raw data records get transformed into structured data records. That might be unexpected for some destinations.